pipeline.go

```diff
@@ -200,57 +229,84 @@ func NewPipelineDefinition(path string, env *PipelineEnvironment) (*PipelineDefi
 	}
 	d := &PipelineDefinition{}
 	err = yaml.Unmarshal(data, d)
-	// Add default meta to services and steps
-	for name, s := range d.Services {
-		s.Meta.Init()
-		d.Services[name] = s
-	}
-	for name, s := range d.Steps {
-		s.Meta.Init()
-		s.Meta.KeepAlive = KeepAliveNo
-		d.Steps[name] = s
-	}
 	// Update with specific meta if defined
-	for name, meta := range env.Services {
-		s, ok := d.Services[name]
-		if ok {
-			s.Meta = meta
-			d.Services[name] = s
-		} else if !meta.Ignore {
-			log.Printf("Metadata: unknown service '%s'", name)
-		}
-	}
 	for name, meta := range env.Steps {
 		s, ok := d.Steps[name]
 		if ok {
+			meta.Type = s.Meta.Type
 			s.Meta = meta
-			s.Meta.KeepAlive = KeepAliveNo
+			if meta.Type == ServiceTypeStep {
+				s.Meta.KeepAlive = KeepAliveNo
+			}
 			d.Steps[name] = s
 		} else if !meta.Ignore {
 			log.Printf("Metadata: unknown step '%s'", name)
 		}
 	}
+	// Initialize all metadata
+	for n, step := range d.Steps {
+		step.Meta.Init()
+		d.Steps[n] = step
+	}
 	return d, err
 }
 
 // Pipelines calculates and verifies dependencies and ordering for steps
 // defined in the PipelineDefinition p.
 func (p *PipelineDefinition) Pipelines() (*Pipelines, error) {
 	if p.pipelines == nil {
-		steps := make(map[string]Step, 0)
-		// Collect ignored steps names
+		// Collect ignored and selected steps names
 		ignoredSteps := types.StringSet{}
+		selectedSteps := types.StringSet{}
 		for name, step := range p.Steps {
 			if step.Meta.Ignore {
 				ignoredSteps[name] = true
 			}
+			if step.Meta.Selected {
+				selectedSteps[name] = true
+				if step.Meta.Ignore {
+					return nil, fmt.Errorf("Instructed to ignore selected step '%s'", step.Name)
+				}
+			}
 		}
-		for name, step := range p.Services {
-			if step.Meta.Ignore {
+
+		// If steps or services are marked es selected, expand the selection
+		queue := make([]string, 0)
+		for name := range selectedSteps {
+			queue = append(queue, name)
+		}
+		if len(queue) > 0 {
```

Comment:
This `if` is redundant, the `for` below just won't execute. If you need it for the update step I would add it only there instead, this also removes at least one level of code block depth. That said I don't think you need the check for the Update either as you iterate over `p.Steps` below anyway

Reply:
Ok, I thought about that while I was writing it and was unsure where to put it. The if is only needed for the update logic after the for-loop, but the for-loop consumes the whole queue and thus I'd need to add another variable to show that the action needs to be performed.